File: ll_stack.c
/* Implementing the stack data structure using linked list in C */ 

/* Strategy */ 
/* 1. Define node using struct 
   2. Define stack 
   3. Initialize stack
   4. Check if the stack is empty
   5. Create a new node on the stack
   6. Push a new node onto the stack
   7. Pop the top node from the stack
   8. Peek at the top node of the stack
   9. Print the stack
*/ 

#include <endian.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Node data structure 
typedef struct Node {
	int data;
	struct Node *next;
} Node;

// Define stack
typedef struct Stack {
	Node *head;
} Stack;

// Initialize the stack (the default stack will be empty which means the head will point to null
void initialize_stack(Stack *stack) {
	stack->head = NULL;
}

// Check if the stack is empty
bool is_empty(Stack *stack) {
	return stack->head == NULL;
}

// Create a new node
Node* create_new_node(int data) {

	// Memory allocation on the heap for a new node
	Node *new_node = (Node *)malloc(sizeof(Node));
	new_node->data = data;
	return new_node;
}

// Push a new node onto the stack
void push_node(Stack *stack, int data) {

	// Create new node by calling function defined above
	Node *new_node = create_new_node(data);

	// Make sure the memory has bee allocated properl and no mistakes have occurred during this process
	if (!new_node) {
		printf("Memory allocation error");
		return;
	}	

	// Connect the newly created node to the head
	new_node->next = stack->head;

       // Rearrange the pointer to the new head (new node)
       stack->head = new_node;
       
       // Print the new node added at the top of the stack
       printf("Pushed %d\n", stack->head->data );
}

// Pop the top node from the stack
void pop_node(Stack *stack) {

	// Check if the stack is empty
	if (is_empty(stack)) {
		printf("The stack is empty\n");
		return;
	} else {
		// Create a temporary node pointing to the head
		Node *temp = stack->head;

		// Shift the head to the next node
		stack->head = stack->head->next;

		// Print the node we're popping off of the stack
		printf("Popped %d\t\n", temp->data);
		
		// Free memory occupied by the temp node
		free(temp);
	}
}

// Peek at the top element of the stack
void peek_element(Stack *stack) {

	// Check if the stack is empty
	if (is_empty(stack)) {
		printf("The stack is empty\n");
		return;
	} else {
		printf("The top element %d\t\n", stack->head->data);
	}
}

// Print elements of the stack
void print_stack(Stack *stack) {

	// Create a temp node to traverse through the linked list and print data on each iteration
	// starting from the top of the stack
	printf("Stack: ");
	Node *temp = stack->head;
	while (temp) {
		printf("%d\t", temp->data );
		temp = temp->next;
	}	
	printf("\n");
}

/* Main function */
int main(void) {

// Create a new stack dynamically on the heap at runtime
Stack *st = (Stack *)malloc(sizeof(Stack));

// Initialize the stack
initialize_stack(st);
// Push onto the stack
push_node(st, 10);
push_node(st, 20);
push_node(st, 30);
push_node(st, 40);
push_node(st, 50);
print_stack(st);
peek_element(st);
pop_node(st);
pop_node(st);
pop_node(st);
pop_node(st);
pop_node(st);
pop_node(st);
print_stack(st);
free(st);
return 0;

}

File: ll_stack.py
# Implementing stack data structure in Python using OOP linked lists

# Strategy
# 1. Create a node class
# 2. Create a stack class
# 3. In that stack class define methods
# 4. Create an empty stack (initialize the stack using constructor
# 5. Check if the stack is empty
# 6. Push a new node to the stack
# 7. Peek at the top node of the stack
# 8. Pop the top node fron the stack
# 9. Print the stack

# Node class
class Node:

    # Initialize a new node when the class is being instantiated
    def __init__(self, data: int) -> None:
        self.data = data
        self.next = None

# Stack class
class Stack:

    # Initialize an empty stack to get started with
    def __init__(self) -> None:
        self.head = None

    # Methods

    # Check if the stack is empty
    def is_empty(self) -> bool:
        return self.head is None

    # Push a new node to the stack
    def push_node(self, new_data: int) -> object:

        # Allocate memory on the heap for a new node, call the Node constructor
        new_node = Node(new_data)

        # If the memory allocation failed, return
        if (not new_node):
            print(f"Memory allocation failed")
            return

        # Link the new node to the head
        new_node.next = self.head

        # Rearrange the pointer of the head so that it's pointing to the new node
        self.head = new_node
        return new_node

    # Pop the top node from the linked list (stack)
    def pop_node(self):
        
        # Check if the stack is empty
        if (self.is_empty()):
            print("The stack is empty")
            return
            
        # Create a temp variable to point to the head of the stack
        temp = self.head
        print(f"Popped {temp.data}")
        # Shift the head to the next node
        self.head = self.head.next

        # Free memory occupied by temp (the former head)
        del temp
    
    # Peek at the top element
    def peek_top(self):

        # Check if the stack is empty
        if (self.is_empty()):
            print("The stack is empty")
            return

        # Print the top node
        print(f"The top node {self.head.data}")


    # Print the stack
    def print_stack(self):
        temp = self.head
        print("The stack", end=" ")
        while (temp):
            print(temp.data, end=" ")
            temp = temp.next
        print()

if __name__ == '__main__':
    # Instantiate a new stack object
    st = Stack()
    st.push_node(10)
    st.push_node(20)
    st.push_node(30)
    st.push_node(40)
    st.push_node(50)
    st.print_stack()
    st.pop_node()
    st.print_stack()
    st.peek_top()
    st.pop_node()
    st.pop_node()
    st.peek_top() 
    st.print_stack()
